src/wp-includes/blocks.php

```diff
@@ -402,34 +418,20 @@ function register_block_type_from_metadata( $file_or_folder, $args = array() ) {
 	 * instead of reading a JSON file per-block, and then decoding from JSON to PHP.
 	 * Using a static variable ensures that the metadata is only read once per request.
 	 */
-	static $core_blocks_meta;
-	if ( ! $core_blocks_meta ) {
-		$core_blocks_meta = require ABSPATH . WPINC . '/blocks/blocks-json.php';
-	}
 
 	$metadata_file = ( ! str_ends_with( $file_or_folder, 'block.json' ) ) ?
 		trailingslashit( $file_or_folder ) . 'block.json' :
 		$file_or_folder;
 
-	$is_core_block = str_starts_with( $file_or_folder, ABSPATH . WPINC );
-	// If the block is not a core block, the metadata file must exist.
-	$metadata_file_exists = $is_core_block || file_exists( $metadata_file );
-	if ( ! $metadata_file_exists && empty( $args['name'] ) ) {
-		return false;
-	}
-
-	// Try to get metadata from the static cache for core blocks.
-	$metadata = array();
-	if ( $is_core_block ) {
-		$core_block_name = str_replace( ABSPATH . WPINC . '/blocks/', '', $file_or_folder );
-		if ( ! empty( $core_blocks_meta[ $core_block_name ] ) ) {
-			$metadata = $core_blocks_meta[ $core_block_name ];
-		}
-	}
+	$metadata_file_exists = file_exists( $metadata_file );
```

Comment:
This one is tricky because when reading the metadata from the registry, putting a file on the disk is rather redundant. However, for some reason we still keep these files for core blocks. So, we can keep it as a requirement for now, but I would look further later how to refactor code so it wasn't necessary.

Old code:

```php
// If the block is not a core block, the metadata file must exist.
$metadata_file_exists = $is_core_block || file_exists( $metadata_file );
```

Core block folder:

<img width="303" alt="Screenshot 2024-09-26 at 21 17 26" src="https://github.com/user-attachments/assets/37443a4b-7406-453a-9406-0afcf992225e">

Reply:
I found that the existing code wants to know if the file exists so it can support the `render` and `variation` keys. I think this is mostly for themes?

```php
$metadata['file'] = $metadata_file_exists ? wp_normalize_path( realpath( $metadata_file ) ) : null;
```
then:
```php
	if ( ! empty( $metadata['render'] ) ) {
		$template_path = wp_normalize_path(
			realpath(
				dirname( $metadata['file'] ) . '/' .                          // <--- metadata file
				remove_block_asset_path_prefix( $metadata['render'] )
			)
		);
```
and:
```php
	if ( ! empty( $metadata['variations'] ) && is_string( $metadata['variations'] ) ) {
		$variations_path = wp_normalize_path(
			realpath(
				dirname( $metadata['file'] ) . '/' .             // <--- metadata file
				remove_block_asset_path_prefix( $metadata['variations'] )
			)
		);
```

It's also checked when doing translations:
```
			if ( $metadata_file_exists && $textdomain && isset( $i18n_schema->$key ) ) {
				$settings[ $mapped_key ] = translate_settings_using_i18n_schema( $i18n_schema->$key, $settings[ $key ], $textdomain );
			}
```

Reply:
Let's leave it for later, but I think we could see if we can use the folder name instead. If the block needs `render` which is a file like `render.php` or `variations` which is a file like `variations.php` they need to live in some folder and the path is relative, example:

https://github.com/WordPress/wordpress-develop/blob/ad1fabe41719f853ec5cdbdde5a9fba97444ff03/tests/phpunit/tests/blocks/register.php#L967-L977

https://github.com/WordPress/wordpress-develop/blob/bbec266c74fc47ce2919faa4cf3c6d43044588cb/tests/phpunit/data/blocks/notice/block.json#L75

<img width="323" alt="Screenshot 2024-09-26 at 21 37 04" src="https://github.com/user-attachments/assets/7c1cc761-3c16-4ca0-a918-5dde55111793">

Essentially `file:render.php` means, relative to where `block.json` lives, so that could be checked against the physical folder to make all the PHP function calls work if there is no `block.json` on disk.